cmd/traffic/cmd/manager/internal/ambassadoragent/cloudtoken/service.go

```diff
@@ -0,0 +1,121 @@
+package cloudtoken
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/datawire/dlib/dlog"
+	"github.com/datawire/k8sapi/pkg/k8sapi"
+	"github.com/telepresenceio/telepresence/v2/cmd/traffic/cmd/manager/managerutil"
+	apiv1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+type Service interface {
+	MaybeAddToken(ctx context.Context, apikey string) error
+}
+
+type createConfigmapIfNotPresent struct {
+	createToken func(ctx context.Context, apikey string) error
+	done        <-chan struct{}
+}
+
+var (
+	suffix               = "agent-cloud-token"
+	cloudConnectTokenKey = "CLOUD_CONNECT_TOKEN"
+)
+
+// NewCreateConfigmapIfNotPresent will start up watchers to watch for a configmap or secret
+// with the apikey. If one is found, the watchers shut down. If MaybeAddToken is called
+// and successfully creates a configmap, the watchers are shut down.
+func NewCreateConfigmapIfNotPresent(ctx context.Context) *createConfigmapIfNotPresent {
```

Comment:
This function contains three fairly well contained and separate things. There's creation of the watcher, the search function, and the watcherloop. Keeping those three together make this function complex and harder to read. You end up with names like `cancelCtx`and `subCtx` because they are cancelled for different reasons.